compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1582,6 +1582,17 @@ fn detect_self_contained_mingw(sess: &Session) -> bool {
 /// We only provide such support for a very limited number of targets.
 fn self_contained(sess: &Session, crate_type: CrateType) -> bool {
     if let Some(self_contained) = sess.opts.cg.link_self_contained {
+        if self_contained {
+            let supported = match (&*sess.target.os, &*sess.target.env) {
+                ("windows", "gnu") => true,
+                ("linux", "musl") => true,
+                ("wasi", _) => true,
+                _ => false,
+            };
+            if !supported {
```

Comment:
```suggestion
            if sess.target.link_self_contained == LinkSelfContainedDefault::False {
```
I think you can currently just use this condition instead of specifying the targets explicitly.

Also it makes sense to prohibit `-Clink-self-contained=false` too for unsupported targets.

Reply:
Checking for `LinkSelfContainedDefault::False` would currently exclude all `wasm32` targets, not just the `wasi` environment. Though, if we are already using self-contained link on all wasm targets by default, I guess that would mean it's supported on all wasm targets? If so, I should update the docs to show that.

In which scenario would we want to prohibit `-Clink-self-contained=false`?

Reply:
>In which scenario would we want to prohibit `-Clink-self-contained=false`?

`-Clink-self-contained` may become more fine-grained in the future (https://github.com/rust-lang/rust/pull/96884) and be supported for targets for which it's not nominally supported right now but for which we still ship some components (like libunwind for some targets).
So it's better to restrict it now if that's possible.

Reply:
>Checking for `LinkSelfContainedDefault::False` would currently exclude all `wasm32` targets, not just the `wasi` environment.

Could you move `link_self_contained: LinkSelfContainedDefault::True` from `compiler\rustc_target\src\spec\wasm_base.rs` to `compiler\rustc_target\src\spec\wasm32_wasi.rs` then?
(Next to the other self contained config options.)

That would be closer to the truth.